src/shell/layer/mod.rs

```diff
@@ -0,0 +1,489 @@
+    }
+
+    pub fn set_anchor(&self, conn: &mut ConnectionHandle, anchor: Anchor) {
+        match self.kind {
+            // We currently rely on the bitsets being the same
+            LayerKind::Wlr(ref wlr) => wlr
+                .set_anchor(conn, zwlr_layer_surface_v1::Anchor::from_bits_truncate(anchor.bits())),
+        }
+    }
+
+    pub fn set_exclusive_zone(&self, conn: &mut ConnectionHandle, zone: i32) {
+        match self.kind {
+            LayerKind::Wlr(ref wlr) => wlr.set_exclusive_zone(conn, zone),
+        }
+    }
+
+    pub fn set_margin(
+        &self,
+        conn: &mut ConnectionHandle,
+        top: i32,
+        right: i32,
+        bottom: i32,
+        left: i32,
+    ) {
+        match self.kind {
+            LayerKind::Wlr(ref wlr) => wlr.set_margin(conn, top, right, bottom, left),
+        }
+    }
+
+    pub fn set_keyboard_interactivity(
+        &self,
+        conn: &mut ConnectionHandle,
+        value: KeyboardInteractivity,
+    ) {
+        match self.kind {
+            LayerKind::Wlr(ref wlr) => wlr.set_keyboard_interactivity(conn, value.into()),
+        }
+    }
+
+    pub fn set_zdepth(&self, conn: &mut ConnectionHandle, depth: ZDepth) {
+        match self.kind {
+            LayerKind::Wlr(ref wlr) => wlr.set_layer(conn, depth.into()),
+        }
+    }
+
+    pub fn kind(&self) -> &LayerKind {
+        &self.kind
+    }
+
+    pub fn wl_surface(&self) -> &wl_surface::WlSurface {
+        &self.wl_surface
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq)]
+pub enum LayerKind {
+    Wlr(zwlr_layer_surface_v1::ZwlrLayerSurfaceV1),
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum KeyboardInteractivity {
+    /// No keyboard focus is possible.
+    ///
+    /// This is the default value for all newly created layer shells.
+    None,
+
+    /// Request exclusive keyboard focus if the layer is above shell surfaces.
+    ///
+    /// For [`ZDepth::Top`] and [`ZDepth::Overlay`], the seat will always give exclusive access to the layer
+    /// which has this interactivity mode set.
+    ///
+    /// This setting is intended for applications that need to ensure they receive all keyboard events, such
+    /// as a lock screen or a password prompt.
+    Exclusive,
+
+    /// The compositor should focus and unfocus this surface by the user in an implementation specific manner.
+    ///
+    /// Compositors may use their normal mechanisms to manage keyboard focus between layers and regular
+    /// desktop surfaces.
+    ///
+    /// This setting is intended for applications which allow keyboard interaction.  
+    OnDemand,
+}
+
+impl Default for KeyboardInteractivity {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+/// The z-depth of a layer.
+///
+/// These values indicate which order a [`Layer`] is rendered in.
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum ZDepth {
```

Comment:
This, on the other hand, represents the layer, so this enum should be named `Layer`.